### PR TITLE
Widget Demo Client - Initial Version

### PR DESCRIPTION
**Pull request by shundhammer**
## Trello

https://trello.com/c/08xsH3CQ/2407-5-new-installation-client-for-testing-the-ui-theme

## Bugzilla

https://bugzilla.opensuse.org/show_bug.cgi?id=1186426

## Description

This is a new YaST client to showcase most of the UI widgets and for testing widget style sheets. It is a full-fledged YaST client with unit tests and multiple wizard pages (with full "next" / "back" functionality), some with tabs to switch between.

You can start it with or without the wizard steps in the left side panel and with any of the available .qss widget style sheets.

More details in the repo's [README.md](https://github.com/shundhammer/yast-widget-demo/blob/huha-initial/README.md).

### Selecting a .qss Style Sheet

Since the Qt widget style sheet is loaded and applied when the Qt UI is initialized, unfortunately there is no way to let the user select a style sheet as a first step. The style sheet editor (Ctrl-Shift-Alt-S) does this from within the Qt UI; but there is no access to that from the outside.

So the user really has to use the `Y2STYLE` environment variable to use a different one than the default `style.qss` style sheet.

### Wizard Steps Panel or Not?

During installation, this is determined by a rather complex mechanism using the `control.xml` file. openSUSE products use the steps panel, SLE products don't. There is no reasonably simple way to find that out in every environment, especially not in the installed system.

So the user has to invoke the correct client:
- `yast widget_demo` or 
- `yast widget_demo_no_wizard_steps`.

### NCurses Support and Screen Space Limitations

The wizard pages all still fit into the smallest NCurses window that we support: 80x24. The pages fit just barely, and they are not especially pretty in an NCurses window that small, but they fit.

That of course means that we have some limitations how many widgets we can pack into each page. In some of the more crowded pages (in particular the first one: "Overview"), we are at the limit; any widget more will make it necessary to split it up into several pages.

And at a certain number of pages, it becomes painful to get to the page you are interested in (because you just fixed a style sheet problem with a widget that is showcased there). Resist the temptation to add more and more stuff; let's not go overboard with this.

### Widget Options / Modes

Many of our widgets have a number of widget options that change their behaviour and visual appearance. This client can showcase _some_ of them, but not all. For the more exotic ones, please refer to the 256 (!) [widget examples](https://github.com/yast/yast-ycp-ui-bindings/tree/master/examples) that we have.

### Pattern and Package Selector

Simply start the real thing in the installed system. Don't apply any changes, of course. If desired, use a different style sheet as described in this client's [README.md](https://github.com/shundhammer/yast-widget-demo/blob/huha-initial/README.md):

```
sudo Y2STYLE=installation.qss yast2 sw_single
```

Or use one of the [UI examples](https://github.com/yast/yast-ycp-ui-bindings/tree/master/examples). The most comprehensive one is this:

```
/usr/lib/YaST2/bin/y2base ./PatternSelector-wizard-with-packages.rb qt
```

(runs without root permissions)

## No Translations?

No, intentionally no translations. Most of the text is generic "Lorem ipsum" _blind text_. This is to showcase widgets, not to distract or confuse any user with text that might be taken seriously.

## No .desktop File?

No, no .desktop file; intentionally. That would make it show up in the YaST control centers (and in the menus of some of our desktops?) and has a lot of potential to confuse users more than it might help them: Most of the widgets don't do anything, they are filled with generic "Lorem ipsum" text and "Item 01" items, there are (intentionally!) no translations.

If a user simply selects all packages named "yast2-*" from the software selection, he will end up with this one as well, and if it shows up in the control center, we will have to answer the same questions by confused users again and again.

This is a development tool and a demo for (very) interested users. It is not targeted at novice users.

# Screenshots

_see also issue #2_

## Tumbleweed / Qt UI

(Tumbleweed 2021-05-22)

```
yast2 widget_demo
```

widget-demo-tw-01

widget-demo-tw-02

widget-demo-tw-03a

widget-demo-tw-03b

widget-demo-tw-03c

widget-demo-tw-04

widget-demo-tw-05

## SLE-15 SP3 / Qt UI

```
Y2STYLE=sle-installation.qss yast2 widget_demo_no_wizard_steps
```

widget-demo-sle15sp3-01

widget-demo-sle15sp3-02

widget-demo-sle15sp3-03a

widget-demo-sle15sp3-03b

widget-demo-sle15sp3-03c

widget-demo-sle15sp3-04

widget-demo-sle15sp3-05

## NCurses (80x24)

(No difference between Tumbleweed and SLE because there are no widget style sheets for NCurses)

```
yast widget_demo
```

widget-demo-ncurses-01

widget-demo-ncurses-02

widget-demo-ncurses-03a

widget-demo-ncurses-03b

widget-demo-ncurses-03c

widget-demo-ncurses-04

widget-demo-ncurses-05

### CODE REVIEW

**Comment by shundhammer** [comment]
@hellcp: You might be interested in this.

Please notice that this is not about making this demo as pretty as possible, but about showcasing our widgets, and to make .qss widget style sheet development easier.

**Comment by mvidner** [comment]
For usability, I would prefer the left pane to be a (tree) selector for random access of the pages, instead of a Next-Back navigation.

**Comment by shundhammer** [comment]
> For usability, I would prefer the left pane to be a (tree) selector for random access of the pages, instead of a Next-Back navigation.

I understand, but one problem we had in the past was poor visibility of those wizard steps with some style sheets, so presenting them was one goal.

I also thought of adding a wizard menu for direct navigation (additionally to the "Next" and "Back" buttons) and for some other debugging functions, but that would have clashed with the look and feel during installation and even more with the menu bar on one of the pages. Plus, it's not available for NCurses (just like the Wizard tree).

This is a compromise, of course. But demonstrating what it looks like in different settings (installation etc.) won over convenience when going through the demo.

I also thought about using the "debug hotkey" Ctrl-Shift-Alt-D only to find out that it's already consumed by starting the Ruby debugger. :-((

**Comment by mvidner** [comment]
> The wizard pages all still fit into the smallest NCurses window that we support: 80x24

OK, but why? This is not the installer where we want to accommodate ancient terminals.

**Comment by shundhammer** [comment]
It's not set in stone. But I recall discussions about that in the past where people complained when it didn't fit.

Plus, it's also an indication of overcrowding a dialog if it doesn't fit into an 80x24 NCurses window anymore.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    - name: Inspect Installed Packages
+      run: rpm -qa | sort
+
+    - name: Unit Tests
+      run: rake test:unit
+      # enable code coverage reporting
+      env:
+        COVERAGE: 1
+
+    ### # send the coverage report to coveralls.io
```
Comment (mvidner) [inline .github/workflows/ci.yml]: let's remove commented out code

File: .rubocop.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  Max: 105
+
+Metrics/BlockLength:
+  # rspec is known as DSL with big blocks
+  Exclude:
+    - test/**/*
+
+# some storage API have size method, but without empty? method
+# for details see https://github.com/yast/yast-storage-ng/pull/83
+Style/ZeroLengthPredicate:
```
Comment (mvidner) [inline .rubocop.yml]: I have checked that this can be removed

File: .rubocop.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+# this is failing in storage wrapper, it will need more time to adapt
+Style/FrozenStringLiteralComment:
+  Enabled: false
+
+# workaround https://github.com/rubocop-hq/rubocop/issues/6861
+Layout/AccessModifierIndentation:
+  EnforcedStyle: indent
+
+Naming/UncommunicativeMethodParamName:
```
Comment (mvidner) [inline .rubocop.yml]: safe to remove this

**Review by mvidner — APPROVED** [review]
Thanks!

1. Just some opportunities to delete leftover meta-code.

2. The trello card says "We need to see all widgets in all states (checked/unchecked checkbox, disabled checked/unchecked checkbox = 4 combinations)." Plus there are tri-state checkboxes too. Wouldn't it be appropriate to do these in this PR?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (shundhammer) [reply .github/workflows/ci.yml]: OK; done.

File: .rubocop.yml (hunk 2 above)
Comment (shundhammer) [reply .rubocop.yml]: OK; done.

File: .rubocop.yml (hunk 3 above)
Comment (shundhammer) [reply .rubocop.yml]: OK; done.

**Comment by shundhammer** [comment]
> The trello card says "We need to see all widgets in all states (checked/unchecked checkbox, disabled checked/unchecked checkbox = 4 combinations)." Plus there are tri-state checkboxes too. Wouldn't it be appropriate to do these in this PR?

For CheckBox true/false/nil (tri-state) and RadioButton in both checked and uncecked as well as enabled and disabled state, that means 6+4 cases, i.e. 12 total. That's too much for the first page.

But you have a point; I think I'll add another page at the very end for this purpose. But in a separate PR.

**Pull request merged**